src/translate/pddl_parser/parsing_functions.py

```diff
@@ -1,31 +1,142 @@
+        assert self._traceback.pop() == message
+
+
+def check_named_block(alist, names):
+    return isinstance(alist, list) and alist and alist[0] in names
+
+def assert_named_block(context, alist, names):
+    if not check_named_block(alist, names):
+        context.expected_named_block_error(alist, names)
+
+def construct_typed_object(context, name, _type):
+    with context.layer("Parsing typed object"):
+        if not isinstance(name, str):
+            context.expected_word_error("Name of typed object", name)
+        return pddl.TypedObject(name, _type)
+
+
+def construct_type(context, curr_type, base_type):
+    with context.layer("Parsing PDDL type"):
+        if not isinstance(curr_type, str):
+            context.expected_word_error("PDDL type", curr_type)
+        if not isinstance(base_type, str):
+            context.expected_word_error("Base type", base_type)
+        return pddl.Type(curr_type, base_type)
+
+
+def parse_typed_list(context, alist, only_variables=False,
+                     constructor=construct_typed_object,
                      default_type="object"):
-    result = []
-    while alist:
-        try:
-            separator_position = alist.index("-")
-        except ValueError:
-            items = alist
-            _type = default_type
-            alist = []
-        else:
-            items = alist[:separator_position]
-            _type = alist[separator_position + 1]
-            alist = alist[separator_position + 2:]
-        for item in items:
-            assert not only_variables or item.startswith("?"), \
-                   "Expected item to be a variable: %s in (%s)" % (
-                item, " ".join(items))
-            entry = constructor(item, _type)
-            result.append(entry)
-    return result
+    with context.layer("Parsing typed list"):
+        result = []
+        group_number = 1
+        while alist:
+            with context.layer(f"Parsing {group_number}. group of typed list"):
+                try:
+                    separator_position = alist.index(TYPED_LIST_SEPARATOR)
+                except ValueError:
+                    items = alist
+                    _type = default_type
+                    alist = []
+                else:
+                    if separator_position == len(alist) - 1:
+                        context.error(
+                            f"Type missing after '{TYPED_LIST_SEPARATOR}'.",
+                            alist)
+                    items = alist[:separator_position]
+                    _type = alist[separator_position + 1]
+                    alist = alist[separator_position + 2:]
+                    if not (isinstance(_type, str) or
+                           (_type and _type[0] == "either" and
+                            all(isinstance(_sub_type, str) for _sub_type in _type[1:]))):
+                        context.error("Type value is expected to be a single word "
+                                      "or '(either WORD*)")
```

Comment:
This message was at first unclear to me. Maybe expand it to the following?

Type value is expected to be a single word, or a choice of words denoted by '(either word1 word2 ...)'

Reply:
Now from my digging in the issue tracker yesterday it sounded like we don't support `either`. Here I see the opposite and from the code it seems it's supported. We might need to clean that up in the issue tracker then.

That being said, I prefer your formulation.